SharedCode/src/commonMain/kotlin/ApplicationPresenter.kt

```diff
@@ -67,11 +65,39 @@ class ApplicationPresenter: ApplicationContract.Presenter() {
     }
 
     override fun setDepartureStation(departureStation: String) {
-        chosenDepartureStation = departureStation
+        chosenDepartureStation = matchStationNameToCode(departureStation)
     }
 
     override fun setArrivalStation(arrivalStation: String) {
-        chosenArrivalStation = arrivalStation
+        chosenArrivalStation = matchStationNameToCode(arrivalStation)
+    }
+
+    private fun buildStationDropdowns() {
+        launch {
+            val allStationDetails = getAllStationDetails()
+            buildStationNamesList(allStationDetails)
+            view!!.setDepartureDropdown(stationNames)
+            view!!.setArrivalDropdown(stationNames)
+        }
+    }
+
+    private fun buildStationNamesList(allStationDetails: StationDetails) {
+        for ((codeIndex, stationCode) in stationCodes.withIndex()) {
+            for (stationInformation in allStationDetails.stations) {
+                if (stationInformation.crs == stationCode) {
+                    stationNames[codeIndex] = stationInformation.name
+                }
+            }
+        }
+    }
+
+    private suspend fun getAllStationDetails(): StationDetails {
+        val apiCall = "https://mobile-api-dev.lner.co.uk/v1/stations"
+        return client.get(apiCall)
+    }
+
+    private fun matchStationNameToCode(stationName: String): String {
+        return stationCodes[stationNames.indexOf(stationName)]
```

Comment:
Every time you call this function you're searching the list of station names, which I believe is quite long, can you think of any way you could reduce the amount of you have to traverse over this list?

Reply:
I think I'm only iterating over `stationNames` which is just the small list of station names based on our given station codes (length 5)

Reply:
Ah yes, true. Given that and the fact I've now properly processed how the two station arrays are working, it seems like you're storing two separate arrays of properties about stations, and you're relying on the fact that the elements in the same position in the two arrays correspond to the same actual station. It probably won't go wrong since you've only got five stations at the moment, but it's not a very nice pattern. Can you think of any better way to handle this?

Reply:
I originally set up a map between station codes and station names, but the UI requires a list rather than a collection (which map.values() supplies) so got a little stuck with that... Is there a way to get a list from a map given its not ordered?

Reply:
If you're worried about whether or not those gives you lists of the same order, then you're still relying on having two lists which are in the same order 😛 Rather than continuing to use your original stationCodes list as is later on, is there a way you could produce a single list that has both the codes and the names in?